File: lc/Tree/101. 对称二叉树.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def helper(l, r):
            if not l and not r:
                return True
            
            if not l or not r:
                return False
            
            if l.val != r.val:
                return False
            
            return helper(l.left, r.right) and helper(l.right, r.left)
        
        return helper(root.left, root.right)

# 跟100类似

File: lc/Tree/101. 对称二叉树.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {

    var dfs func (l *TreeNode, r *TreeNode) bool
    dfs = func (l *TreeNode, r *TreeNode) bool{
        if l == nil && r == nil {
            return true
        }
        if l == nil || r == nil {
            return false
        }

        if l.Val != r.Val {
            return false
        }

        return dfs(l.Left, r.Right) && dfs(l.Right, r.Left)
    }

    return dfs(root.Left, root.Right)
}

// ----------------

func isSymmetric(root *TreeNode) bool {
    return helper(root.Left, root.Right)
}

func helper(l,r *TreeNode) bool {
    if l == nil && r == nil {
        return true
    }
    if l == nil || r == nil {
        return false
    }

    if l.Val != r.Val {
        return false
    }
    return helper(l.Left, r.Right) && helper(l.Right, r.Left)
}